README.md
# Flickrie

**NOTICE: This gem is deprecated. You should switch to the
[flickr-objects](https://github.com/janko-m/flickr-objects) gem.**

---

This gem is a nice wrapper for the Flickr API with an object-oriented interface.

- Homepage: [http://janko-m.github.com/flickrie/](http://janko-m.github.com/flickrie/)

- GitHub page: [https://github.com/janko-m/flickrie](https://github.com/janko-m/flickrie)

- Documentation: [http://rubydoc.info/github/janko-m/flickrie/master](http://rubydoc.info/github/janko-m/flickrie/master)

- Wiki: [https://github.com/janko-m/flickrie/wiki](https://github.com/janko-m/flickrie/wiki)

- Flickr API page: [http://www.flickr.com/services/api](http://www.flickr.com/services/api/)

Supported Ruby versions:

- `1.9.2`
- `1.9.3`

This gem follows [Semantic versioning](http://semver.org/).

## Installation and setup

Add it to your `Gemfile`.

```ruby
gem "flickrie"
```

And in your Terminal run `bundle install`.

Now you have to set your API key and shared secret in your app (if you don't have them yet,
you can get them [here](http://www.flickr.com/services/apps/create/apply)).

```ruby
require 'flickrie'

Flickrie.api_key = "API_KEY"
Flickrie.shared_secret = "SHARED_SECRET"
```

If you're in a Rails app, this would go into an initializer.

## Usage

```ruby
set_id = 819234
photos = Flickrie.photos_from_set(set_id) # => [#<Photo: id="8232348", ...>, #<Photo: id="8194318", ...>, ...]

photo = photos.first
photo.id          # => "8232348"
photo.url         # => "http://www.flickr.com/photos/67313352@N04/8232348"
photo.title       # => "Samantha and me"
photo.owner       # => #<User: nsid="67313352@N04", ...>
photo.owner.nsid  # => "67313352@N04"
```

You can also pass in additional parameters to to get more information about photos:

```ruby
photos = Flickrie.photos_from_set(819234, extras: "owner_name,last_update,tags,views")

photo = photos.first
photo.tags.join(" ") # => "cave cold forrest"
photo.owner.username # => "jsmith"
photo.updated_at     # => 2012-04-20 23:29:17 +0200
photo.views_count    # => 24
```

For the list of all Flickr API methods and parameters you can pass to them take a look at the
[Flickr API documentation](http://www.flickr.com/services/api/). The method names in this gem correspond
to Flickr's method names (so, for example, `Flickr.photos_from_set` is actually `flickr.photosets.getPhotos`).

You can also get additional info on a single photo:

```ruby
photo = Flickrie.get_photo_info(8232348)

photo.description           # => "In this photo, Samantha and me found a secret tunnel..."
photo.comments_count        # => 6
photo.visibility.public?    # => true
photo.can_download?         # => true
photo.owner.real_name       # => "John Smith"
photo.location.country.name # => "United States"
```

If you already have an instatiated photo, you can also get info like this:

```ruby
photo.description # => nil
photo.get_info
photo.description # => "In this photo Peter said something really funny..."
```

You'll also probably want to display these photos in your app. There are
some neat methods to help you with this.

```ruby
photo = Flickrie.get_photo_sizes(8232348)
# or `photo.get_sizes` on an existing photo

photo.medium!(500)
photo.size       # => "Medium 500"
photo.source_url # => "http://farm8.staticflickr.com/7090/7093101501_9337f28800.jpg"
photo.width      # => 375
photo.height     # => 500

photo.available_sizes # => ["Square 75", "Thumbnail", "Square 150", "Small 240", "Small 320", "Medium 500"]
```

So, in your ERB template you could do something like this (in Rails):

```erb
<%= image_tag photo.source_url, :size => "#{photo.width}x#{photo.height}" %>
```

You can see the full list of available methods and attributes in the
[documentation](http://rubydoc.info/gems/flickrie/).

Also, be sure to check the [wiki](https://github.com/janko-m/flickrie/wiki) for some additional info and tips.

## Authentication

```ruby
require 'flickrie'

Flickrie.api_key = "API_KEY"
Flickrie.shared_secret = "SHARED_SECRET"

request_token = Flickrie::OAuth.get_request_token
url = request_token.authorize_url
puts "Visit this url to authenticate: #{url}"

print "If you agreed, the code was displayed afterwards. Enter it: "
code = gets.strip
access_token = request_token.get_access_token(code)
puts "You successfully authenticated!"

Flickrie.access_token = access_token.token
Flickrie.access_secret = access_token.secret
access_token.user_params # => {fullname: "John Smith", user_nsid: "67131352@N03", username: "jsmith"}
```

When getting the authorization url, you can also call
```ruby
request_token.authorize_url(permissions: "read")
```
to ask only for "read" permissions. Available permissions are "read", "write" and "delete".

If you want to make authentication in your web application, I would highly
recommend the [omniauth-flickr](https://github.com/timbreitkreutz/omniauth-flickr) gem.
Or, if you want to do it from scratch, without any additional gems, check out
[this wiki](https://github.com/janko-m/flickrie/wiki/Authentication-in-web-applications)
for an example how that authentication might look like.

## Photo upload

```ruby
photo_id = Flickrie.upload("/path/to/photo.jpg", title: "A cow")
photo = Flickrie.get_photo_info(photo_id)
photo.title # => "A cow"
```

For the list of parameters you can pass in when uploading a photo, see
[this page](http://www.flickr.com/services/api/upload.api.html).

Note that photo uploads require authentication with "write" permissions.

See [this wiki](https://github.com/janko-m/flickrie/wiki/Asynchronous-photo-upload) for an example
of an asynchronous photo upload.

## A few words

Now, there are a lot more API methods that I didn't cover yet,
but I'll constantly update this gem with new API methods. For all of the methods
I didn't cover, you can still call them using

```ruby
Flickrie.client.get(method_name, params = {})
Flickrie.client.post(method_name, params = {})
```

For example:

```ruby
response = Flickrie.client.get "flickr.photos.getContext", :photo_id => 2842732

reponse.body # =>
# {
#   "count" => {"_content" => 99},
#   "prevphoto" => {
#     "id" => "6946978706",
#     "secret" => "b38270bbd6",
#     ...
#   }
#   "nextphoto" => {
#     "id" => "6946979704",
#     "secret" => "74513ff732",
#     ...
#   }
# }
response.body['prevphoto']['id'] # => "6946978706"
```

It's not nearly as pretty, but at least you can get to the data for the
time being. Notice that the `:api_key` parameter is always passed in by
default.

## Credits

Special thanks to @**mislav**, my brother, he helped me really a lot
with getting started with Ruby, and with the basis of this gem.

## Changelog

You can see the changelog [here](https://github.com/janko-m/flickrie/blob/master/CHANGELOG.md).

## Social

You can follow me on Twitter, I'm [@jankomarohnic](https://twitter.com/jankomarohnic).

## License

[MIT](https://github.com/janko-m/flickrie/blob/master/LICENSE)


CHANGELOG.md
# Flickrie changelog

## Version 1.6.1

- Update to SSL endpoint (Flickr deprecated "http").

## Version 1.6.0

- Cover some more API methods, finishing the part with sets:

```
flickr.photosets.setPrimaryPhoto
flickr.photosets.reorderPhotos
flickr.photosets.removePhoto
flickr.photosets.removePhotos
flickr.photosets.orderSets
flickr.photosets.getContext
flickr.photosets.editMeta
flickr.photosets.delete
flickr.photosets.create
flickr.photosets.editPhotos
flickr.photosets.addPhoto
```

## Version 1.5.2

- Fixed a bug when error would be raised if a certain (unimportant)
  attribute wouldn't exist.

## Version 1.5.1

- Fixed a bug where methods like `Flickrie::User#get_info` were still
  using the global access token instead of the local one belonging to
  a specific `Flickrie::Instance`.

## Version 1.5.0

- When you're using an API method for fetching photos which has the
  `:extras` parameter, instead of passing
  `:extras => 'url_sq,url_t,url_q,url_s,url_n,url_m,url_z,url_c,url_l,url_h,url_k,url_o'`,
  you can now just pass `:include_sizes => true`, and that will do the
  same thing.

- Covered `flickr.photos.transform.rotate` and `flickr.photos.licenses.setLicense`.

- New homepage for the project. Check it out:
  [http://janko-m.github.com/flickrie/](http://janko-m.github.com/flickrie/).

## Version 1.4.2

- Update documentation

## Version 1.4.1

- Covered a bunch more methods, making the "photos" and "people" sections
  practically finished:

```
flickr.people.getPhotosOf
flickr.photos.getPerms
flickr.photos.getRecent
flickr.photos.getUntagged
flickr.photos.getWithGeoData
flickr.photos.getWithoutGeoData
flickr.photos.recentlyUpdated
flickr.photos.setContentType
flickr.photos.setDates
flickr.photos.setMeta
flickr.photos.setPerms
flickr.photos.setSafetyLevel
flickr.photos.setTags
```

## Version 1.4.0

- Accidentally released `1.4.1` instead of this one, so there is no
  `1.4.0` :P

## Version 1.3.1

- Fix `Flickrie::Collection` using `will_paginate` if available,
  not only when it's set with `Flickrie.pagination`.

- Fix some collections still being a simple `Array`, instead of
  `Flickrie::Collection`.

## Version 1.3.0

- Implemented short URLs for photos and videos (`#short_url`).

- When fetching multiple objects (can be photos, sets, users, or
  whatever), if now returns an instance of `Flickrie::Collection`.
  This is basically an `Array` with a few extra features. See
  [`Flickrie::Collection`](http://rubydoc.info/gems/flickrie/Flickrie/Collection)
  for more details.

- Implemented pagination. Sorry it took so long. But it's really cool,
  you can even use [will_paginate](https://github.com/mislav/will_paginate) with it.
  See [`Flickrie.pagination`](http://rubydoc.info/gems/flickrie/Flickrie#pagination-class_method)
  for more details.

- Covered `flickr.photos.getNotInSet` and `flickr.people.getUploadStatus`.

## Version 1.2.0

- Added 2 new Flickr sizes (`Large 1600` and `Large 2048`).

- The width and height is now back to normal again, Flickr didn't
  mix them up, I did (when a photo is rotated, its width and height
  stay the same). Sorry.

## Version 1.1.2

- Fixed documentation not displaying properly.

## Version 1.1.1

- Excluded unnecessary files from the release (1.4MB => 48KB).
  Sorry :P

## Version 1.1.0

- Before `Flickrie::User` had only the `#public_photos` method, now, I
  just added the missing `#public_videos`, `#public_media`, `#photos`,
  `#videos` and `#media`.

- Fixed Flickr's bug with mixing up the width and the height of a photo
  (not only in the API, but also on their website).

## Version 1.0.2

- Fix some mistakes in the documentation

## Version 1.0.1

- Transfered the documentation to YARD. [Documentation](http://rubydoc.info/gems/flickrie/)

## Version 1.0.0

### Changes that are backwards compatible

- Fixed content type not being passed to the file you're uploading
   if you specified it directly.

- The request will now be retried once more if it timeouts

- In authentication:

 - you can now call `request_token.authorize_url` instead of `request_token.get_authorization_url`.

 - you can now call `request_token.get_access_token(code)` instead of `Flickrie::OAuth.get_access_token(code, request_token)`.

 - you also get the infomation about the user who just authenticated,
   which you can then access with `access_token.user_info`
   (it's a Hash with keys `:fullname`, `:user_nsid` and `:username`)

- When calling `Flickrie.get_photos_counts`, the `Flickrie::MediaCount`
  now also has attributes `#time_interval` (alias for `#date_range`),
  `#from` and `#to`.

### Changes that are NOT backwards compatible

- If you're passing in the content type of the file you're uploading,
  the parameter is now called `:content_type`, not `:mime_type`.

- If there is a problem in obtaining the access token, `Flickrie::Error`
  is now raised, instead of `Flickrie::OAuth::Error`.

- When you're calling `request_token.get_authorization_url`, if you want to
  specifiy permissions, you now have to pass the `:perms` option,
  instead of `:permissions`. In this way you can pass any parameter,
  and it will be appended to the URL (in case Flickr adds a new parameter).

- When you have a `Flickrie::User` instance, the
  `Flickrie::User#time_zone` now returns a struct with `#label` and
  `#offset` attributes (before it returned a `Hash` with those keys).

- When you call `Flickrie.get_media_context`, the result is now a
  struct with attributes `#count`, `#previous`, `#next`.

## Version 0.7.3

- Covered `flickr.people.getPhotos` (I released it as a patch because
  it's important).

## Version 0.7.2

- Specified versions of dependencies and Ruby accurately.

## Version 0.7.1

- Fixed an oauth bug.
- Added `#hash` method to photos, sets etc. which you can call to get the
  raw data (in Ruby hash). This is useful if I accidentally left something out,
  or if Flickr added something new that I didn't cover yet, that you can

lib/photo.rb
require 'object'

module Flickr
  class Photo < Flickr::Object
    SIZES = {
      'Square 75'  => 'sq',
      'Square 150' => 'q',
      'Thumbnail'  => 't',
      'Small 240'  => 's',
      'Small 320'  => 'n',
      'Medium 500' => 'm',
      'Medium 640' => 'z',
      'Medium 800' => 'c',
      'Large 1024' => 'l',
      'Original'   => 'o'
    }

    attr_reader :size

    def id
      @hash['id'].to_i
    end

    def title
      @hash['title']
    end

    def available_sizes
      SIZES.select { |_, size_abbr| @hash["url_#{size_abbr}"] }.keys
    end

    def square(number)
      Photo.new(@hash, "Square #{number}")
    end

    def thumbnail
      Photo.new(@hash, "Thumbnail")
    end

    def small(number)
      Photo.new(@hash, "Small #{number}")
    end

    def medium(number)
      Photo.new(@hash, "Medium #{number}")
    end

    def large(number)
      Photo.new(@hash, "Large #{number}")
    end

    def original
      Photo.new(@hash, "Original")
    end

    def largest
      Photo.new(@hash, largest_size)
    end

    def width
      @hash["width_#{size_abbr}"].to_i
    end

    def height
      @hash["height_#{size_abbr}"].to_i
    end

    def url
      @hash["url_#{size_abbr}"]
    end

    def to_s
      url
    end

    def inspect
      attributes = %w[size url width height].inject({}) do |hash, attr|
        hash.update(attr => send(attr))
      end
      %(#<Photo: #{attributes.map { |k, v| %(#{k}="#{v}") }.join(", ")}>)
    end

    class InvalidSize < ArgumentError
    end

    private

    def initialize(hash, size = nil)
      @hash = hash
      @size = size || largest_size

      unless available_sizes.include?(@size)
        raise InvalidSize, "size \"#{@size}\" isn't available for this photo"
      end
    end

    def size_abbr
      SIZES[@size]
    end

    def largest_size
      available_sizes.last
    end
  end
end

__END__

{
  "id" => "6913731566",
  "secret" => "23879c079a",
  "server" => "7130",
  "farm" => 8,
  "dateuploaded" => "1333956611",
  "isfavorite" => 0,
  "license" => "0",
  "safety_level" => "0",
  "rotation" => 0,
  "owner" => {
    "nsid" => "67131352@N04",
    "username" => "Janko Marohnić",
    "realname" => "",
    "location" => "",
    "iconserver" => "0",
    "iconfarm" => 0
  },
  "title" => {"_content" => "6913664138_61ffb9c0d7_b"},
  "description" => {"_content" => ""},
  "visibility" => {"ispublic" => 1, "isfriend" => 0, "isfamily" => 0},
  "dates" => {
    "posted" => "1333956611",
    "taken" => "2012-04-09 00:30:11",
    "takengranularity" => "0",
    "lastupdate" => "1333956616"
  },
  "views" => "0",
  "editability" => {"cancomment" => 0, "canaddmeta" => 0},
  "publiceditability" => {"cancomment" => 1, "canaddmeta" => 0},
  "usage" => {"candownload" => 1, "canblog" => 0, "canprint" => 0, "canshare" => 0},
  "comments" => {"_content" => "0"},
  "notes" => {"note" => []},
  "people" => {"haspeople" => 0},
  "tags" => {"tag" => []},
  "urls "=> {
    "url"=> [
      {
        "type"=>"photopage",
       "_content"=>"http://www.flickr.com/photos/67131352@N04/6913731566/"
      }
    ]
  },
  "media"=>"photo"}
}
